encoding/kzgrs/prover/prover.go

```diff
@@ -0,0 +1,280 @@
+	*kzgrs.KzgConfig
+	Srs          *kzg.SRS
+	G2Trailing   []bls.G2Point
+	mu           sync.Mutex
+	LoadG2Points bool
+
+	ParametrizedProvers map[rs.EncodingParams]*ParametrizedProver
+}
+
+func NewProver(config *kzgrs.KzgConfig, loadG2Points bool) (*Prover, error) {
+
+	if config.SRSNumberToLoad > config.SRSOrder {
+		return nil, errors.New("SRSOrder is less than srsNumberToLoad")
+	}
+
+	// read the whole order, and treat it as entire SRS for low degree proof
+	s1, err := kzgrs.ReadG1Points(config.G1Path, config.SRSNumberToLoad, config.NumWorker)
+	if err != nil {
+		log.Println("failed to read G1 points", err)
+		return nil, err
+	}
+
+	s2 := make([]bls.G2Point, 0)
+	g2Trailing := make([]bls.G2Point, 0)
+
+	// PreloadEncoder is by default not used by operator node, PreloadEncoder
+	if loadG2Points {
+		if len(config.G2Path) == 0 {
+			return nil, fmt.Errorf("G2Path is empty. However, object needs to load G2Points")
+		}
+
+		s2, err = kzgrs.ReadG2Points(config.G2Path, config.SRSNumberToLoad, config.NumWorker)
+		if err != nil {
+			log.Println("failed to read G2 points", err)
+			return nil, err
+		}
+
+		g2Trailing, err = kzgrs.ReadG2PointSection(
+			config.G2Path,
+			config.SRSOrder-config.SRSNumberToLoad,
+			config.SRSOrder, // last exclusive
+			config.NumWorker,
+		)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// todo, there are better ways to handle it
+		if len(config.G2PowerOf2Path) == 0 {
+			return nil, fmt.Errorf("G2PowerOf2Path is empty. However, object needs to load G2Points")
+		}
+	}
+
+	srs, err := kzg.NewSrs(s1, s2)
+	if err != nil {
+		log.Println("Could not create srs", err)
+		return nil, err
+	}
+
+	fmt.Println("numthread", runtime.GOMAXPROCS(0))
+
+	encoderGroup := &Prover{
+		KzgConfig:           config,
+		Srs:                 srs,
+		G2Trailing:          g2Trailing,
+		ParametrizedProvers: make(map[rs.EncodingParams]*ParametrizedProver),
+		LoadG2Points:        loadG2Points,
+	}
+
+	if config.PreloadEncoder {
+		// create table dir if not exist
+		err := os.MkdirAll(config.CacheDir, os.ModePerm)
+		if err != nil {
+			log.Println("Cannot make CacheDir", err)
+			return nil, err
+		}
+
+		err = encoderGroup.PreloadAllEncoders()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return encoderGroup, nil
+
+}
+
+func (g *Prover) PreloadAllEncoders() error {
+	paramsAll, err := GetAllPrecomputedSrsMap(g.CacheDir)
+	if err != nil {
+		return nil
+	}
+	fmt.Printf("detect %v srs maps\n", len(paramsAll))
+	for i := 0; i < len(paramsAll); i++ {
+		fmt.Printf(" %v. NumChunks: %v   ChunkLen: %v\n", i, paramsAll[i].NumChunks, paramsAll[i].ChunkLen)
+	}
+
+	if len(paramsAll) == 0 {
+		return nil
+	}
+
+	for _, params := range paramsAll {
+		// get those encoders and store them
+		enc, err := g.GetKzgEncoder(params)
+		if err != nil {
+			return err
+		}
+		g.ParametrizedProvers[params] = enc
+	}
+
+	return nil
+}
+
+func (g *Prover) GetKzgEncoder(params rs.EncodingParams) (*ParametrizedProver, error) {
```

Comment:
`GetKzgProver`? Returning a prover while naming the method as "encoder" would be confusing.

Reply:
The final PR will focus on code cleanup, and I will address this there. Right now I'm focused on the high level file restructuring.